docs(framework) Add user authentication guide (#5630)
adap/flower: framework/docs/source/how-to-login-to-superlink.rst

Login to SuperLink
==================

.. note::

    This guide describes a Flower Enterprise feature. Subscribe to the `Flower
    Enterprise plan <https://flower.ai/enterprise>`_ to get access to the advanced
    features for your organization.

In this guide, you'll learn how to configure SuperLink with user-level authentication

> and authorization, and how to log in to the SuperLink using the ``flwr`` CLI. Once

Reviewer: ```suggestion
and authorization, and how to log using the ``flwr`` CLI. Once
```

It sounds a bit strange to say `log in to the SuperLink`

logged in, any users that are authorized on the SuperLink can run Flower CLI commands
that interact with the SuperLink.

.. important::

    With user authentication and authorization enabled, only users that have submitted
    the ``flwr run`` command can view and interact with their runs. This means that your
    runs are **private by default**, ensuring that only authorized users can access
    them.

Prerequisites
-------------

To enable user authentication and authorization, the SuperLink must be deployed with an
`OIDC <https://openid.net/developers/how-connect-works/>`_ provider and an `OpenFGA
<https://openfga.dev/>`_ server. The OIDC provider is used for user authentication,
while OpenFGA is used for fine-grained access control. This means an authenticated user
can only run ``flwr`` CLI commands on the SuperLink if they have been granted the
necessary permissions by the SuperLink administrator.

.. note::

    When enabled, both user authentication and authorization must be configured on the

>     SuperLink.

Reviewer: do we need this note?

Author: We should have this sentence, but doesn't have to be a note. I'll revise it.

    authorization:
      auth_type: openfga
      authz_url:          # The base OpenFGA API URL
      store_id:           # The store ID containing the model store
      model_id:           # The model ID containing the latest authorization model for the SuperLink
      relation:           # The authorized relation between the user and the resource, e.g. `has_access`
      object:             # The object and identifier at which a user has an authorized relation.
                          # The expected format is `object_type:object_identifier`, e.g.:
                          #   service:researchgrid
                          #   │       └─ Identifier of the object
                          #   └─ Object type

Save this file as ``user-auth-config.yaml``. Then pass it to the SuperLink via the
``--user-auth-config`` flag when deploying the SuperLink:

.. code-block:: bash

    ➜ flower-superlink \
        --user-auth-config=user-auth-config.yaml
        <other flags>

.. note::

    To authorize a user, the SuperLink administrator must add the user's OIDC ``sub``
    claim to the OpenFGA store with the appropriate relation.

Login to the SuperLink
----------------------

> Once a SuperLink with user authentication and authorization is up and running, you can
> interface with it by installing the ``flwr`` PyPI package. Then, ensure that the
> ``enable-user-auth`` field is set to ``true`` in the federation section in the
> ``pyproject.toml`` of your Flower app:

Reviewer: ```suggestion
Once a SuperLink with user authentication and authorization is up and running, a user can
interface with it after installing the ``flwr`` PyPI package via the Flower CLI. Then, ensure that the
``enable-user-auth`` field is set to ``true`` in the federation section in the
``pyproject.toml`` of the Flower app you want to run:
```

.. code-block:: toml

    [tool.flwr.federations]
    default = "my-federation"

    [tool.flwr.federations.my-federation]
    address = "<SUPERLINK-ADDRESS>:9093"   # Address of the SuperLink Exec API
    root-certificates = "<PATH/TO/ca.crt>" # TLS certificate set for the SuperLink. Required for self-signed certificates.
    enable-user-auth = true                # Enables the user auth mechanism on the `flwr` CLI side

.. note::

    User authentication and authorization is only supported with TLS connections.

Now, you need to login first before other CLI commands can be executed. Upon executing
``flwr login``, a URL will be returned by the authentication plugin in the SuperLink.
Click on it and authenticate directly against the OIDC provider.

.. code-block:: bash

    flwr login [APP] [FEDERATION]
    Loading project configuration...
    Success
    Creating a new `.gitignore` with `.credentials` entry...

>     Please login with your user credentials here: https://account.flower.ai/realms/flower/device?user_code=DRVT-YQMA

Reviewer: ```suggestion
    Please login with your user credentials here: https://account.flower.ai/realms/flower/device?user_code=...
```

    # [... user follows URL and logs in ... in the meantime the CLI will wait ...]
    ✅ Login successful.

Once the login is successful, the credentials returned by the OIDC provider via the
SuperLink will be saved to the app's directory under
``.flwr/.credentials/<federation-name>.json``. The tokens stored in this file will be
sent transparently with each subsequent ``flwr`` CLI request to the SuperLink, and it
will relay them to OIDC provider to perform the authentication checks.

Run authorized ``flwr`` CLI commands
------------------------------------

> With the above steps completed, you can now run ``flwr`` CLI commands that require
> authentication and authorization. For example, as an authorized user, you can run the

Reviewer: ```suggestion
With the above steps completed, you can now run ``flwr`` CLI commands against a SuperLink setup with user
authentication and authorization. For example, as an authorized user, you can run the
```
